### PR TITLE
Add support for loading and parsing an ELF kernel

### PR DESCRIPTION
**Pull request by conradgrobler**
This change adds support for Loading an ELF kernel from stage0 using the QEMU fw_cfg device, parsing it, and laying it out in memory.

QEMU assumes that an ELF kernel passed using the `-kernel` command-line argument will use the PVH boot protocol and lays it out into the guest memory. This causes a problem when booting with SEV enabled, since the kernel data in guest memory does not get encrypted. To work around it we pass the raw kernel ELF file manually via the fw_cfg device.

QEMU does not accept the `-append` or `-initrd` arguments if `-kernel` is not specified, so we also pass those in manually via the fw_cfg device.

Parsing an ELF kernel required adding a global allocator, since paring ELF program headers return a vector.

I also had to enable LTO to make sure stage0 still fits within the expected size with all the additional code added.

### CODE REVIEW

**Review by ipetr0v — APPROVED** [review]

File: stage0/.cargo/config.toml [hunk 1]
```diff
@@ -2,9 +2,9 @@
 target = "x86_64-unknown-none"
 
 [target.x86_64-unknown-none]
-rustflags = "-C relocation-model=static -C code-model=large"
+rustflags = "-C relocation-model=static -C code-model=large -C linker-plugin-lto"
```
Comment (andrisaar) [inline stage0/.cargo/config.toml:5]: Is adding `lto = true` to `Cargo.toml` not enough? We have that for the release builds, but not for dev builds.

If the dev build size is the problem, we could also just bump up the binary size, as LTO makes it more difficult to debug things.

File: stage0/.cargo/config.toml [hunk 2]
```diff
@@ -2,9 +2,9 @@
...
 
 [unstable]
 # We need to build std / core because of the code model. This may be removed if / when it's possible to build stage0 without it in the future.
-build-std = ["core"]
+build-std = ["core", "alloc"]
```
Comment (andrisaar) [inline stage0/.cargo/config.toml:9]: We managed to get so far without this :(

File: stage0/src/kernel.rs [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+/// QEMU fw_cfg device.
+const DEFAULT_KERNEL_START: u64 = 0x200000;
+
+/// The default size for the kernel if a kernel wasn't supplied via the QEMU fw_cfg device.
+///
+/// This is an arbitrary value, just to ensure it is non-zero.
+const DEFAULT_KERNEL_SIZE: usize = 0x100000;
+
+/// The file path used by Stage0 to read the kernel from the fw_cfg device.
+const KERNEL_FILE_PATH: &[u8] = b"opt/stage0/elf_kernel\0";
```
Comment (andrisaar) [inline stage0/src/kernel.rs:37]: I feel like this should be in `kernel.rs` similar to all the other paths (like for the initramfs or arguments).

Comment (andrisaar) [reply stage0/src/kernel.rs:37]: Oops, that is actually here.

File: stage0/src/main.rs [hunk 4]
```diff
@@ -63,12 +67,12 @@ static SEV_SECRETS: MaybeUninit<oak_sev_guest::secrets::SecretsPage> = MaybeUnin
...
 static SEV_CPUID: MaybeUninit<oak_sev_guest::cpuid::CpuidPage> = MaybeUninit::uninit();
 
-/// The default entry point for the kernel if one wasn't supplied via the QEMU fw_cfg device.
-const DEFAULT_KERNEL_ENTRY: u64 = 0x200000;
-
 /// We create an identity map for the first 1GiB of memory.
 const TOP_OF_VIRTUAL_MEMORY: u64 = Size1GiB::SIZE;
 
+/// The file path used by Stage0 to read the kernel command-line from the fw_cfg device.
+const CMDLINE_FILE_PATH: &[u8] = b"opt/stage0/cmdline\0";
```
Comment (andrisaar) [inline stage0/src/main.rs:74]: I think the cmdline stuff should be in `kernel.rs` as well.

**Review by andrisaar — APPROVED** [review]

File: stage0/.cargo/config.toml (hunk 1 above)
Comment (conradgrobler) [reply stage0/.cargo/config.toml:5]: It wasn't previously when I was fully parsing the ELF file, but it now works again with parsing only the header and the program headers. Reverted this change.

File: stage0/.cargo/config.toml (hunk 2 above)
Comment (conradgrobler) [reply stage0/.cargo/config.toml:9]: Indeed.

File: stage0/src/main.rs (hunk 4 above)
Comment (conradgrobler) [reply stage0/src/main.rs:74]: Moved the code.

**Review by conradgrobler — COMMENTED** [review]
Thanks

**Pull request merged**